src/python/pants/engine/target.py

```diff
@@ -1570,9 +1570,65 @@ class DependenciesRequest:
     field: Dependencies
 
 
+@union
+@dataclass(frozen=True)
+class InjectDependenciesRequest(ABC):
+    """A request to inject dependencies, in addition to those explicitly provided.
+
+    To set up a new injection, subclass this class. Set the class property `inject_for` to the
+    type of `Dependencies` field you want to inject for, such as `FortranDependencies`. This will
+    cause the class, and any subclass, to have the injection. Register this subclass with
+    `UnionRule(InjectDependenciesRequest, InjectFortranDependencies)`, for example.
+
+    Then, create a rule that takes the subclass as a parameter and returns `InjectedDependencies`.
+
+    For example:
+
+        class FortranDependencies(Dependencies):
+            pass
+
+        class InjectFortranDependencies(InjectDependenciesRequest):
+            inject_for = FortranDependencies
+
+        @rule
+        def inject_fortran_dependencies(request: InjectFortranDependencies) -> InjectedDependencies:
+            return InjectedDependencies([Address.parse("//:injected")]
+
+        def rules():
+            return [
+                inject_fortran_dependencies,
+                UnionRule(InjectDependenciesRequest, InjectFortranDependencies),
+            ]
+    """
+
+    field: Dependencies
+    inject_for: ClassVar[Type[Dependencies]]
+
+
+class InjectedDependencies(DeduplicatedCollection[Address]):
+    sort_input = True
+
+
 @rule
-def resolve_dependencies(request: DependenciesRequest) -> Addresses:
-    return Addresses(request.field.sanitized_raw_value or ())
+async def resolve_dependencies(
+    request: DependenciesRequest, union_membership: UnionMembership
+) -> Addresses:
+    provided = request.field.sanitized_raw_value or ()
+
+    # Inject any dependencies. This is determined by the `request.field` class. For example, if
+    # there is a rule to inject for FortranDependencies, then FortranDependencies and any subclass
+    # of FortranDependencies will use that rule.
+    inject_request_types = cast(
+        Iterable[Type[InjectDependenciesRequest]],
+        union_membership.union_rules.get(InjectDependenciesRequest, ()),
```

Comment:
Can this method be made generic to avoid the need to cast at callsites?

Reply:
Indeed! In a followup, I plan to add some methods to `UnionMembership` to make it more ergonomic, given how widely we use it now.